Add internal annotation

giacomocavalieri opened the pull request:
This PR adds the `@internal` annotation, closing #2486 

### ~~Why is this a WIP PR?~~

- [x] I'm waiting for #2537 to be merged and exclude internals from the package interface as well

DanielSherlock commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -368,6 +368,9 @@ where
 
         // Module types
         for (name, type_) in &module.ast.type_info.types {
+            if type_.internal {
+                continue;
+            }
```
Are these lines of code excluding `@internal` types from the same file from completions? I don't think we want that.

giacomocavalieri replied:
I think you're right! I will look into it
Thanks for pointing it out!!

giacomocavalieri replied:
I checked and you're perfectly right! Thank you so much for the help 💜

DanielSherlock commented on compiler-core/src/language_server/tests/completion.rs:
```diff
@@ -726,3 +726,71 @@ pub fn wibble(
+}
+
+#[test]
+fn public_internal_definitions_from_the_same_module_are_in_the_completions() {
```
This test seems to only check value-completion, not type-completions. In particular the types `Foo` and `Alias`.

I think these would need a separate test (like `for_custom_type_definition()` at line 568 or the five following tests) which would then detect the I-think-incorrect behaviour in my above comment.

DanielSherlock commented:
_(I don't rust, so feel free to ignore me completely.)_

I joined the stream late so was curious to look at parts of the code again after, and noticed some things that seem suspicious about completions for types. Hopefully the comments I added are getting at the right bits of code for the issue.

giacomocavalieri commented:
Thank you @DanielSherlock for pointing this out! I've fixed the bug and added some tests to make sure that internal types do appear in the completions for the same module they're defined in

lpil commented:
Could you rebase on main please? 🙏

giacomocavalieri commented on compiler-core/src/analyse.rs:
```diff
@@ -537,7 +547,14 @@ fn register_value_from_function(
+        *public,
+        *internal,
+        deprecation.clone(),
+    );
```
Ok! Should it be a compile error or should it be a bit more permissive and the formatter could just remove the internal annotation on private functions?

lpil replied:
Oh yes good point. Perhaps a compile error would be good?

giacomocavalieri replied:
Ok! I went with an error

giacomocavalieri commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -382,7 +385,7 @@ where
             // Qualified types
             for (name, type_) in &module.types {
-                if !type_.public {
+                if !type_.public || type_.internal {
```
Wouldn't removing it show _all_ internal types in the autocompletions - even those defined by external dependencies?

lpil replied:
Aye, so it needs to check the package too

giacomocavalieri replied:
I'm not sure how I can check if a module come from the root package, how can I do that?

lpil commented:
Is this WIP @giacomocavalieri ?

giacomocavalieri commented:
Yes I still have to finish addressing your review comments, I'll be working on this shortly!

giacomocavalieri commented:
Turning `public` + `internal` into a single `publicity` attribute turned out to be quite the chunky refactoring but I think this is finally ready for review @lpil 🎉

lpil commented on compiler-core/src/docs.rs:
```diff
@@ -164,6 +167,7 @@ pub fn generate_html(
             .ast
             .definitions
             .iter()
+            .filter(|statement| !statement.is_internal())
```
nit: is_public() would be a wee bit clearer here

giacomocavalieri replied:
Aha the problem is that we want to do it with both public and internals, so anything that is not private, how would you call such a method?
```rust
fn wibble(&self) -> bool {
  case self {
    Publicity::Internal | Publicity::Public => true,
    Publicity::Private => false,
  }
}
```

lpil replied:
I realised this for the later ones 😓 

I wonder if `!x.is_internal()` should be `x.is_importable()` or something?

giacomocavalieri replied:
Ah I really like that name, I'll go for it and we can always change it later 😁

lpil commented on compiler-core/src/docs.rs:
```diff
@@ -559,7 +563,7 @@ fn type_<'a>(source_links: &SourceLinker, statement: &'a TypedDefinition) -> Opt
 
     match statement {
-        Definition::CustomType(ct) if ct.public && !ct.opaque => Some(Type {
+        Definition::CustomType(ct) if !ct.publicity.is_private() && !ct.opaque => Some(Type {
```
Same here: `.is_public()`

lpil commented on compiler-core/src/javascript/typescript.rs:
```diff
@@ -294,41 +294,43 @@ impl<'a> TypeScriptGenerator<'a> {
                 type_,
                 ..
-            }) if *public => vec![self.type_alias(alias, type_)],
+            }) if !publicity.is_internal() => vec![self.type_alias(alias, type_)],
```
Why is this rendered for public and private but not internal?

giacomocavalieri replied:
Oh that's a typo! It was meant to be `!publicity.is_private()`

lpil replied:
oh no! 😁

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -377,10 +388,13 @@ where
+                        completions.push(type_completion(None, unqualified.used_name(), type_))
+                    }
+                    None => continue,
+                }
```
In this case they've been imported explicitly so I think we should show them.

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -354,6 +355,8 @@ where
             completions.push(type_completion(None, name, type_));
         }
 
+        let root_package = self.compiler.project_compiler.config.name.clone();
```
I don't think we need to clone here, using a `&str` will be fine. A helper function `self.root_package_name()` could be good too

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -392,9 +406,14 @@ where
             completions.push(value_completion(None, name, value));
         }
 
+        let root_package = self.compiler.project_compiler.config.name.clone();
```
Same as above

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -418,10 +445,13 @@ where
-                };
-                completions.push(value_completion(None, unqualified.used_name(), value));
+                match module.get_public_value(&unqualified.name) {
+                    Some(value) if value.publicity.is_internal() => continue,
```
As above, they've been explicitly brought into scope so let's autocomplete them

lpil commented on compiler-core/src/language_server/tests/completion.rs:
```diff
@@ -19,13 +23,33 @@ fn positioned_expression_completions(
+/// `external` is used to tell whether the specified dependency has to be
+/// treated as coming from an external imported package or as coming from a
+/// module from the same root package.
+/// In any case `src` will be able to import it using `import dep`.
```
External means it's written in FFI, dependency means it comes from another package. I think with the old name it should have always been in a dep package.

giacomocavalieri replied:
The way it was implemented it was just considered as another module under the same root package. I can change it to make sure the dep is always under another package if you want

lpil commented on compiler-core/src/language_server/tests/completion.rs:
```diff
@@ -19,13 +23,33 @@ fn positioned_expression_completions(
     _ = io.src_module("app", src);
+
     let response = engine.compile_please();
+    println!("{:#?}", response);
```
Printing here 😁

giacomocavalieri replied:
🫢

lpil reviewed (commented):
Nice!! And what an effort 💪  I've left a few small notes inline.

lpil reviewed (approved):
Thank you!!!
